src/python/pants/backend/java/util_rules.py

```diff
@@ -3,79 +3,119 @@
 
 from __future__ import annotations
 
+import textwrap
 from dataclasses import dataclass
+from typing import ClassVar, Iterable
 
 from pants.backend.java.compile.javac_subsystem import JavacSubsystem
-from pants.engine.fs import Digest
+from pants.engine.fs import CreateDigest, Digest, FileContent, FileDigest, MergeDigests
 from pants.engine.internals.selectors import Get
-from pants.engine.process import FallibleProcessResult, Process, ProcessCacheScope, ProcessResult
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessCacheScope
 from pants.engine.rules import collect_rules, rule
+from pants.jvm.resolve.coursier_fetch import (
+    Coordinate,
+    Coordinates,
+    CoursierLockfileEntry,
+    ResolvedClasspathEntry,
+)
 from pants.jvm.resolve.coursier_setup import Coursier
 
 
 @dataclass(frozen=True)
 class JdkSetup:
     digest: Digest
-    java_home_cmd: tuple[str, ...]
-    fingerprint_comment: str
+    nailgun_jar: str
+    jdk_preparation_script: ClassVar[str] = "__jdk.sh"
+    java_home: ClassVar[str] = "__java_home"
```

Comment:
Why is this not a constant?

Reply:
`ClassVars` are effectively constant unless we expect this to be subclassed. But the real answer is that I was just following the existing pattern in the `JavacBinary` class.